### PR TITLE
[nodejs] add base images

### PR DESCRIPTION
**Pull request by rochdev**
## Motivation

Avoid redoing this on every commit.

## Changes

Add base images for Node.js

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

**Comment by rochdev** [comment]
@cbeauchesne Do you know why the images are not building? I copied everything from Python and yet it doesn't look to be doing anything.

File: utils/build/docker/nodejs/nextjs.Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM node:20-slim
+FROM datadog/system-tests:nextjs.base-v0
 
 RUN apt-get update && apt-get install -y jq curl git
```
Comment (cbeauchesne) [inline utils/build/docker/nodejs/nextjs.Dockerfile:3]: Did you forgot to remove this ?

File: utils/build/docker/nodejs/uds-express4.Dockerfile [hunk 2]
```diff
@@ -1,4 +1,4 @@
-FROM node:18-slim
+FROM datadog/system-tests:express4.base-v0
 
 RUN apt-get update && apt-get install -y jq curl git
```
Comment (cbeauchesne) [inline utils/build/docker/nodejs/uds-express4.Dockerfile:3]: Same comment

File: utils/build/docker/nodejs_otel/express4-otel.Dockerfile [hunk 3]
```diff
@@ -1,6 +1,6 @@
-FROM node:18
+FROM datadog/system-tests:express4.base-v0
 
-RUN apt-get update && apt-get install -y jq
+RUN apt-get update && apt-get install -y jq curl git
```
Comment (cbeauchesne) [inline utils/build/docker/nodejs_otel/express4-otel.Dockerfile:3]: Same comment

File: utils/build/build_nodejs_base_images.sh [hunk 4]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+# build and push nodejs base images
+
+
+docker buildx build --load --progress=plain -f utils/build/docker/nodejs/express4.base.Dockerfile -t datadog/system-tests:express4.base-v0 .
+docker buildx build --load --progress=plain -f utils/build/docker/nodejs/nextjs.Dockerfile -t datadog/system-tests:nextjs.base-v0 .
```
Comment (cbeauchesne) [inline utils/build/build_nodejs_base_images.sh:7]:
```suggestion
docker buildx build --load --progress=plain -f utils/build/docker/nodejs/nextjs.base.Dockerfile -t datadog/system-tests:nextjs.base-v0 .
```

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by cbeauchesne** [comment]
Right now, there is a small egg-chicken bug introduced in my last PR on pull command. Once this PR is ready, if I didn't fixed it, I'll hack it.

**Comment by rochdev** [comment]
I made the changes requested, so in theory this should be good to go, but I don't know how to test it and adding the label doesn't seem to have worked.

**Comment by cbeauchesne** [comment]
You have an error in your script : https://github.com/DataDog/system-tests/actions/runs/9912836061/job/27388583055#step:5:409

File: utils/build/build_nodejs_base_images.sh [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+
+# build and push nodejs base images
```
Comment (cbeauchesne) [inline utils/build/build_nodejs_base_images.sh:3]: Don't let the script fail silently : 

```suggestion
# build and push nodejs base images

set -e
```

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Comment by rochdev** [comment]
@cbeauchesne I've merged the latest changes from the main branch and tried fixing the script, but I still don't understand how to make this work. It doesn't seem to be building the base images at all even though the label is on the PR, so every test fails because the base image cannot be found.

**Comment by rochdev** [comment]
Putting this as ready for review since the code is identical to Python, so my assumption is that the failures are because something needs to be done by the R&P team (which seems to also be hinted to in the doc).

File: utils/build/docker/nodejs/express4.base.Dockerfile [hunk 6]
```diff
@@ -0,0 +1,25 @@
+FROM node:18-alpine
+
+RUN apk add --no-cache bash curl git jq
+
+WORKDIR /usr/app
+
+COPY utils/build/docker/nodejs/install_ddtrace.sh binaries* /binaries/
```
Comment (cbeauchesne) [inline utils/build/docker/nodejs/express4.base.Dockerfile:7]: Do not ship anything from system-tests source in base images, it must be added in the build step.

The only thing that can be done is apt-get/apk & co

**Review by cbeauchesne — CHANGES_REQUESTED** [review]
It's failing because the path to install_ddtrace is not the good one. But see my comment, base image must only install dependencies, but must not ship anythong from system-tests sources, as they'll be added during the `./build.sh` phase.

**Comment by cbeauchesne** [comment]
I put back the Pr in draft mode, to save a little bit my notification stack. Put it back to normal when you need a review. 

Side node, theoritically, it requires not manual step from us. But it'll be a good idea to wait for Roberto or myseilf to merge this PR, to handle any issue on gitlab process (it'll be gitlab who will build and push the images that'll be used every where)

**Comment by rochdev** [comment]
> It's failing because the path to install_ddtrace is not the good one.

Now it's the good one and still failing as the base image doesn't seem to be getting built at all.

**Comment by cbeauchesne** [comment]
> the base image doesn't seem to be getting built at all.

They are built : https://github.com/DataDog/system-tests/actions/runs/10099872631/job/27930016319?pr=2677#step:5:123

But it fail because it try to the  get the express base image, while the job is on the nextjs weblog : https://github.com/DataDog/system-tests/actions/runs/10099872631/job/27930016319?pr=2677#step:9:131

**Pull request closed without merging**